add db migration from java

sns-seb opened the pull request:
This PR adds a component that wraps around the ActiveRecord based migration process and handles concurrency as well as asynchronous execution of it

this will be called from the new WS to trigger DB migration (another PR)

Sorry this PR is a little bit dense, but individual commit do not really make sense

sns-seb commented:
rebased on master

simonbrandhof commented on server/sonar-server/src/main/java/org/sonar/server/db/migration/DatabaseMigration.java:
```diff
@@ -0,0 +1,71 @@
+   *
+   * @return a {@link Date} or {@code null}
+   */
+  @Nullable
```
For now we use CheckForNull annotation for method returns and Nullable for method parameters. That was the rule for a correct support by Findbugs. I have to check now the status with latest Java plugin.

simonbrandhof replied:
I confirm that @CheckForNull must be used (see http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)

sns-seb replied:
I should have written in the javadoc "This method should be named {@code start} but it can not be because it will be called by the pico container and will cause unwanted behavior"

`startMigration` has the "migration" word duplicated, it terms of object design, that's not so good. Currently, calling this method can be written `databaseMigration.startIt()`. Since I can not use `start`, I really don't know which is worse of `startIt` and `startMigration` :(

julienlancelot commented on server/sonar-server/src/main/java/org/sonar/server/db/migration/DatabaseMigration.java:
```diff
@@ -0,0 +1,71 @@
+   * <strong>Do not rename this method to {@code start} otherwise it will be called by the pico container</strong>
+   * </p>
+   */
+  void startIt();
```
Shouldn't it better to rename it 'startMigration()' ?

sns-seb replied:
I should have written in the javadoc "This method should be named {@code start} but it can not be because it will be called by the pico container and will cause unwanted behavior"

`startMigration` has the "migration" word duplicated, it terms of object design, that's not so good. Currently, calling this method can be written `databaseMigration.startIt()`. Since I can not use `start`, I really don't know which is worse of `startIt` and `startMigration` :(

julienlancelot replied:
Indeed both are not good, let's keep it like it is

julienlancelot commented on server/sonar-server/src/main/java/org/sonar/server/db/migration/DatabaseMigration.java:
```diff
@@ -0,0 +1,71 @@
+import javax.annotation.Nullable;
+import java.util.Date;
+
+public interface DatabaseMigration {
```
Why have you created an interface ?

sns-seb replied:
"loose coupling in other components"/"mocking in tests"/"best practice"...

why does it surprise you ?

julienlancelot replied:
Because I think that when an interface is not used (only one implementation), it's better to have only the implementation, it reduces the number of classes :)

julienlancelot commented on server/sonar-server/src/main/java/org/sonar/server/platform/AbstractStopableExecutorService.java:
```diff
@@ -0,0 +1,121 @@
+                delegate.shutdownNow(); // Cancel currently executing tasks
+                // Wait a while for tasks to respond to being cancelled
+                if (!delegate.awaitTermination(5, TimeUnit.SECONDS)) {
+                    System.err.println("DBMigration pool did not terminate");
```
Why are you using System.err.println ?

sns-seb replied:
good catch

julienlancelot commented on server/sonar-server/src/main/java/org/sonar/server/platform/ruby/PlatformRackBridge.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Implementation of {@link RackBridge} which get access to the Rack object through the {@link ServletContext}.
+ * <p>
+ * This component implements {@link Startable} so that
```
Missing Javadoc

sns-seb replied:
Actually, some javadoc cleaning was missing. Good catch

sns-seb commented:
rebased on master and took feeback into account

sns-seb commented:
rebased on master
+ fix build instability caused by `AbstractClosableExecutorService`
+ fix incorrect indent in source files

GithubIntegration commented:
SonarQube analysis reported 7 new issues:
* 1 blocking
* 4 major
* 1 minor
* 1 info

simonbrandhof commented:
The new package org.sonar.server.db.migration conflicts with org.sonar.server.db.migrations. Moreover both contain a class named DatabaseMigration.

simonbrandhof commented on server/sonar-server/src/main/java/org/sonar/server/db/migration/DatabaseMigration.java:
```diff
@@ -0,0 +1,72 @@
+   * and this will cause unwanted behavior</strong>
+   * </p>
+   */
+  void startIt();
```
To avoid ambiguity, why not naming the method "migrate" ?

sns-seb replied:
such name introduces a "duplication" which is not very Object oriented `DatabaseMigration.migrate`

subject has been discussed with @julienlancelot here : https://github.com/SonarSource/sonarqube/pull/197#discussion-diff-28219803

simonbrandhof commented on server/sonar-server/src/main/java/org/sonar/server/db/migration/DatabaseMigration.java:
```diff
@@ -0,0 +1,72 @@
+  /**
+   * Current status of the migration.
+   *
+   * @return a {@link Status}
```
Seriously, it returns a Status ? :-) This javadoc can be dropped!

simonbrandhof commented on server/sonar-server/src/main/java/org/sonar/server/db/migration/DatabaseMigration.java:
```diff
@@ -0,0 +1,72 @@
+   * <p>
+   * This value is reset when {@link #startIt()} is called.
+   * </p>
+   * @return a {@link Throwable} or {@code null}
```
this javadoc duplicates the signature too.

simonbrandhof commented on server/sonar-server/src/main/java/org/sonar/server/db/migration/PlatformDatabaseMigration.java:
```diff
@@ -0,0 +1,126 @@
+/**
+ * Handles concurrency to make sure only one DB migration can run at a time.
+ */
+public class PlatformDatabaseMigration implements DatabaseMigration {
```
What's the need for the interface DatabaseMigration if there's only a single implementation ?

sns-seb replied:
point also raised by @julienlancelot (https://github.com/SonarSource/sonarqube/pull/197#discussion-diff-28219923)

basically, out of habit and because it really bugs me to mock a concrete class and it is also more costly => slower tests

simonbrandhof replied:
ok, makes sense.

simonbrandhof commented:
The package org.sonar.server.platform was always badly named IMO. According to our past discussion it should be "system", so the package "ruby" deserves to be directly in org.sonar.server.
Moreover the file package_info.java is missing in this directory.

simonbrandhof commented on server/sonar-server/src/main/java/org/sonar/server/db/migration/PlatformDatabaseMigration.java:
```diff
@@ -0,0 +1,126 @@
+        failureError = null;
+        try {
+          Loggers.get(PlatformDatabaseMigration.class).info("Starting DB Migration at {}", startDate);
+          rubyBridge.databaseMigration().trigger();
```
why RubyBridge does not directly execute migration (for instance "rubyBridge.migrate()") ?

sns-seb replied:
I intended `RubyBridge` to create wrappers around Ruby code which could then be put into pico and this way avoid creating the underlying wrapping code each time.

Did not get that far because it meant to instance classes and wrappers in `ServerComponents` which would have messed the code.

It can be removed.

simonbrandhof replied:
no pb here. You can forget.

sns-seb replied:
In the end, I think I'll leave it as it is as it keeps open the possibility to create the `RubyDatabaseMigration` object returned by `rubyBridge.databaseMigration()` in the constructor of `PlatformDatabaseMigration` and save on all the Ruby script parsing and casting which occurs when `RubyBrigde` methods are called

sns-seb commented:
> The new package org.sonar.server.db.migration conflicts with org.sonar.server.db.migrations. Moreover both contain a class named DatabaseMigration

Wanted to put new `DatabaseMigration` into package `org.sonar.server.db.migration` but a class with the same name already existed in the package so I created a new package.

Existing class `DatabaseMigration` is poorly named imo, should be named `MigrationStep` or something alike.

If I could rename the existing class, I would gladly drop package `org.sonar.server.db.migrations`

simonbrandhof commented:
Ok, go for renaming to `MigrationStep` and to drop `org.sonar.server.db.migrations`

sns-seb commented:
@simonbrandhof  I guess you mean to keep `org.sonar.server.db.migrations` (I did the mistake myself) since it is the legacy package (and is full of stuff) ?

simonbrandhof commented:
exact. The goal is to move your code in package `org.sonar.server.db.migrations`

sns-seb commented:
updated PR :

* moved `ruby` package to `org.sonar.server`
* renamed `DatabaseMigration` to `DatabaseStep` (and `DatabaseMigrations` to `DatabaseSteps`), ran upgrade-sonar-it to check it was ok
* removed package `org.sonar.server.db.migration`
* moved `StoppableExecutorService` to `org.sonar.server.util` (and fixed typo in name)

simonbrandhof commented:
LGTM
